File: 02- February/18- Invert Binary Tree/18- Invert Binary Tree (Ahmed Gamal).exs
# Author: Ahmed Gamal

# Definition for a binary tree node.
#
# defmodule TreeNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           left: TreeNode.t() | nil,
#           right: TreeNode.t() | nil
#         }
#   defstruct val: 0, left: nil, right: nil
# end

defmodule Solution do
  @moduledoc """
  Given the root of a binary tree, invert the tree, and return its root.
  to invert a binary tree means to swap every left node in the tree for its corresponding right node.
  we can do this recursively by swapping the left and right subtrees of a node and then calling the function recursively on the left and right subtrees.
  the base case is when we reach a leaf node, which we can just return nil in this case.
  the recursive case will be excuted before the swap, so it'll swap the subtrees from the bottom up.
  """

  @spec invert_tree(root :: TreeNode.t | nil) :: TreeNode.t | nil

    def invert_tree(nil), do: nil
    def invert_tree(root) do
        # return the root node with the left and right subtrees swapped after calling the function recursively on them
        %TreeNode{
            val: root.val,
            left: invert_tree(root.right),
            right: invert_tree(root.left)
        }
    end
end

File: 02- February/18- Invert Binary Tree/18- Invert Binary Tree (Ahmed Gamal).cpp
// Author: Ahmed Gamal

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// for this problem, we can use recursion to solve it
// we need to swap the left and right subtrees of each node
// we can do this by creating a new node with the same value as the current node and the left and right subtrees swapped (recursively)

class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        // base case is when we reach a leaf node
        if(not root)
            return root;
        
        // return a new node with the same value as the current node and the left and right subtrees swapped
        return new TreeNode(root -> val, invertTree(root -> right), invertTree(root -> left));
    }
};